File: frontend/src/stores/centrosCosto.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { useApiStore } from './api'

export interface CentroCosto {
  id_centro_costo: number
  codigo_centro_costo: string
  nombre_centro_costo: string
  descripcion?: string
  id_responsable?: number
  presupuesto_anual: number
  activo: boolean
  fecha_creacion?: string
  fecha_modificacion?: string
  usuario_creacion?: number
  usuario_modificacion?: number
}

export interface CentroCostoCreate {
  codigo_centro_costo: string
  nombre_centro_costo: string
  descripcion?: string
  id_responsable?: number
  presupuesto_anual?: number
  activo?: boolean
}

export interface CentroCostoUpdate {
  codigo_centro_costo?: string
  nombre_centro_costo?: string
  descripcion?: string
  id_responsable?: number
  presupuesto_anual?: number
  activo?: boolean
}

export const useCentroCostoStore = defineStore('centros-costo', () => {
  const apiStore = useApiStore()

  const centrosCosto = ref<CentroCosto[]>([])
  const isLoading = ref(false)

  const obtenerCentrosCosto = async (params?: {
    skip?: number
    limit?: number
    activo?: boolean
  }): Promise<CentroCosto[]> => {
    try {
      isLoading.value = true
      const queryParams = new URLSearchParams()

      if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
      if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
      if (params?.activo !== undefined) queryParams.append('activo', params.activo === true ? 'true' : 'false')

      const url = `/centros-costo?${queryParams.toString()}`
      const response = await apiStore.get(url)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener centros de costo')
      }
      centrosCosto.value = response.data
      return response.data
    } catch (error) {
      console.error('Error al obtener centros de costo:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  const obtenerCentrosCostoActivos = async (): Promise<CentroCosto[]> => {
    try {
      isLoading.value = true
      const response = await apiStore.get('/centros-costo/activos')
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener centros de costo activos')
      }
      centrosCosto.value = response.data
      return response.data
    } catch (error) {
      console.error('Error al obtener centros de costo activos:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  const buscarCentrosCosto = async (
    searchTerm: string,
    params?: {
      activo?: boolean
    }
  ): Promise<CentroCosto[]> => {
    try {
      isLoading.value = true
      const queryParams = new URLSearchParams()
      queryParams.append('q', searchTerm)

      if (params?.activo !== undefined) queryParams.append('activo', params.activo === true ? 'true' : 'false')

      const response = await apiStore.get(`/centros-costo/buscar?${queryParams.toString()}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al buscar centros de costo')
      }
      return response.data
    } catch (error) {
      console.error('Error al buscar centros de costo:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  const obtenerCentroCosto = async (id: number): Promise<CentroCosto> => {
    try {
      const response = await apiStore.get(`/centros-costo/${id}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al obtener centro de costo')
      }
      return response.data
    } catch (error) {
      console.error('Error al obtener centro de costo:', error)
      throw error
    }
  }

  const crearCentroCosto = async (centroCosto: CentroCostoCreate): Promise<CentroCosto> => {
    try {
      const response = await apiStore.post('/centros-costo', centroCosto)
      if (!response.success) {
        throw new Error(response.error || 'Error al crear centro de costo')
      }
      return response.data
    } catch (error) {
      console.error('Error al crear centro de costo:', error)
      throw error
    }
  }

  const actualizarCentroCosto = async (id: number, centroCosto: CentroCostoUpdate): Promise<CentroCosto> => {
    try {
      const response = await apiStore.put(`/centros-costo/${id}`, centroCosto)
      if (!response.success) {
        throw new Error(response.error || 'Error al actualizar centro de costo')
      }
      return response.data
    } catch (error) {
      console.error('Error al actualizar centro de costo:', error)
      throw error
    }
  }

  const eliminarCentroCosto = async (id: number): Promise<void> => {
    try {
      const response = await apiStore.delete(`/centros-costo/${id}`)
      if (!response.success) {
        throw new Error(response.error || 'Error al eliminar centro de costo')
      }
    } catch (error) {
      console.error('Error al eliminar centro de costo:', error)
      throw error
    }
  }

  return {
    // State
    centrosCosto,
    isLoading,

    // Methods
    obtenerCentrosCosto,
    obtenerCentrosCostoActivos,
    buscarCentrosCosto,
    obtenerCentroCosto,
    crearCentroCosto,
    actualizarCentroCosto,
    eliminarCentroCosto
  }
})

File: backend/app/routes/centros_costo.py
"""
API routes for Centros de Costo
"""
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from database import get_db
from schemas import CentroCostoCreate, CentroCostoUpdate, CentroCostoResponse
from crud import centros_costo_crud

router = APIRouter(
    prefix="/centros-costo",
    tags=["Centros de Costo"]
)


@router.get("", response_model=List[CentroCostoResponse])
def listar_centros_costo(
    skip: int = Query(0, ge=0, description="Número de registros a omitir"),
    limit: int = Query(100, ge=1, le=1000, description="Número máximo de registros a retornar"),
    activo: Optional[bool] = Query(None, description="Filtrar por estado activo"),
    db: Session = Depends(get_db)
):
    """
    Listar todos los centros de costo con paginación opcional
    """
    centros = centros_costo_crud.get_all(db, skip=skip, limit=limit, activo=activo)
    return centros


@router.get("/activos", response_model=List[CentroCostoResponse])
def listar_centros_costo_activos(db: Session = Depends(get_db)):
    """
    Listar todos los centros de costo activos
    """
    centros = centros_costo_crud.get_activos(db)
    return centros


@router.get("/buscar", response_model=List[CentroCostoResponse])
def buscar_centros_costo(
    q: str = Query(..., min_length=1, description="Término de búsqueda"),
    activo: Optional[bool] = Query(None, description="Filtrar por estado activo"),
    db: Session = Depends(get_db)
):
    """
    Buscar centros de costo por código o nombre
    """
    centros = centros_costo_crud.search(db, search_term=q, activo=activo)
    return centros


@router.get("/{id_centro_costo}", response_model=CentroCostoResponse)
def obtener_centro_costo(
    id_centro_costo: int,
    db: Session = Depends(get_db)
):
    """
    Obtener un centro de costo por su ID
    """
    centro = centros_costo_crud.get_by_id(db, id_centro_costo)

    if not centro:
        raise HTTPException(
            status_code=404,
            detail=f"Centro de costo con ID {id_centro_costo} no encontrado"
        )

    return centro


@router.post("", response_model=CentroCostoResponse, status_code=201)
def crear_centro_costo(
    centro_costo: CentroCostoCreate,
    db: Session = Depends(get_db)
):
    """
    Crear un nuevo centro de costo
    """
    # Verificar si el código ya existe
    existing_centro = centros_costo_crud.get_by_codigo(db, centro_costo.codigo_centro_costo)
    if existing_centro:
        raise HTTPException(
            status_code=400,
            detail=f"El código de centro de costo '{centro_costo.codigo_centro_costo}' ya existe"
        )

    # TODO: Obtener usuario_id de la sesión/token
    usuario_id = None

    new_centro = centros_costo_crud.create(db, centro_costo, usuario_id)
    return new_centro


@router.put("/{id_centro_costo}", response_model=CentroCostoResponse)
def actualizar_centro_costo(
    id_centro_costo: int,
    centro_costo: CentroCostoUpdate,
    db: Session = Depends(get_db)
):
    """
    Actualizar un centro de costo existente
    """
    # Si se está actualizando el código, verificar que no exista
    if centro_costo.codigo_centro_costo:
        existing_centro = centros_costo_crud.get_by_codigo(db, centro_costo.codigo_centro_costo)
        if existing_centro and existing_centro.id_centro_costo != id_centro_costo:
            raise HTTPException(
                status_code=400,
                detail=f"El código de centro de costo '{centro_costo.codigo_centro_costo}' ya existe"
            )

    # TODO: Obtener usuario_id de la sesión/token
    usuario_id = None

    updated_centro = centros_costo_crud.update(db, id_centro_costo, centro_costo, usuario_id)

    if not updated_centro:
        raise HTTPException(
            status_code=404,
            detail=f"Centro de costo con ID {id_centro_costo} no encontrado"
        )

    return updated_centro


@router.delete("/{id_centro_costo}", response_model=CentroCostoResponse)
def eliminar_centro_costo(
    id_centro_costo: int,
    db: Session = Depends(get_db)
):
    """
    Desactivar un centro de costo (soft delete)
    """
    # TODO: Obtener usuario_id de la sesión/token
    usuario_id = None

    deleted_centro = centros_costo_crud.delete(db, id_centro_costo, usuario_id)

    if not deleted_centro:
        raise HTTPException(
            status_code=404,
            detail=f"Centro de costo con ID {id_centro_costo} no encontrado"
        )

    return deleted_centro
